src/data-analysis-environments.rst

```diff
@@ -0,0 +1,322 @@
+Entity resolution refers to merging and combining different representations for the same real-world entity.
+For example, a dataset may contain values for "University of Michigan", "U of M", "Umich", "UM".
+They identically refer to one entity and data science workers need to replace all the ambiguous values with the same one.
+It can be difficult to solve by an algorithm but can be easily judged by human knowledge.
+A popular approach for these tasks is crowdsourcing - asking human workers to combine the entities.
+However, this may not be scalable. 
+When the size of the dataset is large, it becomes difficult for human workers to split and coordinate the work.
+
+CrowdER addresses this with a hybrid human-machine approach :cite:`wang:2012`.
+On a high level, it first uses an algorithm to filter and cluster potential values with high probabilities of duplicated entities that need crowd workers to judge.
+Instead of showing crowd works pairwise comparisons, it shows cluster-based values and asks crowd workers to group these candidates.
+This hybrid approach improves both efficiency and accuracy for the result.
+
+.. container:: bib-item
+
+  .. bibliography:: data-analysis-environments.bib
+    :filter: key == 'wang:2012'
+
+Tools for Engineering Features
+==============================
+Finally, the core stage of data science is engineering features, which includes understanding patterns and relations in the data, exploring different representations of the data, and finally extracting insights from the data.
+
+Data Visualization
+------------------
+
+Data visualization can help data science workers intuitively see and explore the trends and patterns in data by creating visual representations of the information.
+Vega_ is a popular declarative language for creating interactive visualizations.
+Users define the interactive elements in JSON format, then Vega renders it using Canvas or SVG.
+In Vega, users often need to provide a specification of the visual elements.
+Vega-Lite_ :cite:`satyanarayan:2017` is a simplified version of Vega.
+Vega-Lite automates the construction of some specifications (e.g., axis, legends, scales).
+Vega-Lite is compiled to Vega and can only express a subset of interactive visualizations in Vega.
+
+ .. _Vega: https://vega.github.io/vega/
+ .. _Vega-Lite: https://vega.github.io/vega-lite/
+
+.. container:: bib-item
+
+  .. bibliography:: data-analysis-environments.bib
+    :filter: key == 'satyanarayan:2017'
+
+Novice users may find it difficult to draw connections with the specification and the runtime state in Vega.
+Hoffswell et al. :cite:`hoffswell:2018` explored the design to visualize program variables within the code editor.
+With this in situ visualization technique, novice users reported performance improvements in both speed and accuracy.
+
+.. image:: https://idl.cs.washington.edu/static/images/figures/insitu-code-vis.png
```

Comment:
Any images you want to include should be included in the repository, since we can't control whether that URL will stay valid